PackageKit/backends/zypp/pk-backend-zypp.cpp

```diff
@@ -4193,7 +4324,12 @@ backend_upgrade_system_thread (PkBackendJob *job, GVariant *params, gpointer use
 	try
 	{
 		pk_backend_job_set_status (job, PK_STATUS_ENUM_REFRESH_CACHE);
-		if (!zypp_refresh_cache (job, zypp, FALSE)) {
+		// Setting force to TRUE, as we want to force a cache refresh
+		// before installing upgrades, now that we use a separate cache,
+		// but we don't want to refresh when doing a complete upgrade.
+		// (only in minimal aka download-only mode)
+		gboolean do_refresh = (parameters->upgrade_kind == PK_UPGRADE_KIND_ENUM_MINIMAL);
```

Comment:
You're accessing parameters here, which was deleted in line 4314. That's why I got so many skipped refreshs before OS download, even though the refresh was forced.